### ISSUE TITLE
Elevated user segregation from public facing instances

### ISSUE THREAD

**Issue by coolio986** [nopSolutions/nopCommerce#7315]
This may or may not have been discussed before, correct me if this is a duplicate request.

I currently use nopCommerce for my ecommerce platform. It works well.

I, however, run my platform slightly different than how nopCommerce was originally designed. I am a C# senior software developer (USA), which enables me to customize this platform the way I prefer. I have been operating my nopCommerce store for several years now, and I wanted to propose a feature with a security mindset.

Issue:
In its current form, nopCommerce allows admin level users to log into the admin console from the same web URL as a customer. This means that there is a potential security vector for bad actors to exploit any zero day attack that may be present in the .net framework. From my time in the industry any elevated access to sensitive information should be separated or firewalled (WAFs) short circuiting any type of exploitation from occurring altogether.

Ideas:
1. Allow two separate instances of nopCommerce (configurable in appsettings.json or any other configuration file at the deployment level). One instance is a customer facing side that does not allow elevated access to the admin control panel. The other instance (elevated) has full control over everything.
2. Elevated access is only accessible from an intranet, or from firewall limits / whitelist.
3. One could argue that any URL with a relative location ie https://yousite/Admin could be managed in a WAFs, but this does not short circuit elevated permissions entirely.
4. Whitelist / blacklist IP's or DNS' from access to the admin site.

What I've done in my fork of nopCommerce:
1. Extended appsettings to allow for elevated permissions
2. Deployed 2 instances (docker) of nopCommerce connected to the same DB. One is customer facing, one is admin facing.
3. Disabled checkout in the admin facing (appsettings option) site as an added precaution to prevent malicious activity.
5. Added a WAFs in front of nopCommerce on both customer and admin facing sites.
6. Have one unified version of nopCommerce for both customer and admin levels.
7. Firewalled the admin site and blacklisted all traffic except from my whitelisted IP's
8. The two instances also shield the customer facing side from any performance loss from admin level activity, keeping the customer site running as smoothly as possible. This also allows the admin site to be an interim dev deployment / testing area before the changes are pushed to the customer / public site.

Cost implications:
It is more expensive to deploy more instances when you are using one instance per server (if using VPS or similar tech for deployment). However, I have been able to deploy all this for about $25 / month which is far cheaper than an AIO solution like Shopify.

These are my ideas, It would be great to hear of any other ideas if nothing like this already exists.

**Comment by danFbach** [comment]
Not sure if you are aware, but there is a setting in `/Admin/Setting/GeneralCommon` under `Security`, `Admin area allowed IP` which allows you to provide a comma separated list of IP's that are allowed to access the admin area, essentially blacklisting everything BUT the provided IP's, which i feel like would provide most of the protection you are requesting here without significantly complicating the codebase and deployment.

**Comment by coolio986** [comment]
@danFbach, Thanks for that. I forgot these setting existed. 
I also forgot to add another issue item above

If an error occurs on the admin site (testing site again) it gets caught before it's rolled out to the customer side. I've seen a few instances where the entire program crashes and takes down the entire site which cannot happen in production.

Maybe my solution is multifaceted, rather than just from a security standpoint. This issues also grazes CI/CD too, not entirely sure how that fits in for others.

**Comment by AndreiMaz** [comment]
Having two instances is not the best way to go because it'll affect your website performance (distributed cache, etc). The easiest and the fastest (in terms of performance) way to go is th suggested point 4 (Whitelist / blacklist IP'sfrom access to the admin site).

You can do it as danFbach suggested. Or use services such as Cloudflare (you can configure these rules on its "Free" plan)

**Comment by coolio986** [comment]
Thanks @AndreiMaz 

I have experienced the performance issue you mentioned. In my environment I am backed by a CDN which improves performance on the customer site even with caching disabled (customer site caching is turned off).

I have been operating in this 2 instance environment for a few years now and while I have experienced a few issues with the caching mechanism, I have been able to mitigate most of those issues in code without degrading performance.

The main reasoning behind this is to completely sever any possibility that a bad actor could gain access to admin level functions.
The other reason is to allow any performance degradation that would otherwise occur on the customer site (by doing admin level functions) to be moved to another instance, unloading the customer site from any admin level functions, keeping the customer site as fast as possible

One argument could be made, using the whitelist and caching would overall be faster anyway and could potentially negate needing a admin instance altogether. My counter to that is regarding testing and crashing for new updates.
The admin site doubles as a development testing ground so any updates can be tested before they are pushed to production. 

This, of course, has nothing to do with caching or performance or whitelisting, but since it is already separated, it makes sense to me.

At this point I am only trying to have a conversation to see if I really should move my entire architecture to a 1 instance configuration. Weighing the merits between both environments. 

Let me know if anyone has any thoughts. If not, I will close this issue and move on.
